foreman/data_refinery_foreman/gatherer/management/commands/gather_accessions.py

```diff
@@ -0,0 +1,731 @@
+            errors.append("The -s, --since date must be earlier than -u, --until date.")
+
+        keyword = options["keyword"]
+        organism = options["organism"]
+        if options["microarray"]:
+            ae_id = options["ae_id"] or options["ae_ids_file"]
+            gpl_id = options["gpl_id"] or options["gpl_ids_file"]
+            ids = ae_id or gpl_id
+            invalid_options_message = (
+                "Exactly one of the keyword [-k, --keyword], organism [-o, --organism] or "
+                "ArrayExpress ID(s) [--ae-id, --ae-ids-file] / GEO platform ID(s) "
+                "[--gpl-id, --gpl-ids-file] must be specified."
+            )
+        elif options["rna_seq"]:
+            taxon_id = options["taxon_id"] or options["taxon_ids_file"]
+            ids = taxon_id
+            invalid_options_message = (
+                "Exactly one of the keyword [-k, --keyword], organism [-o, --organism] "
+                "or taxon ID(s) [--taxon-id, --taxon-ids-file] must be specified."
+            )
+
+        if len([option for option in (ids, keyword, organism) if option]) != 1:
+            errors.append(invalid_options_message)
+
+        if errors:
+            exit("\n".join(errors))
+
+    def handle(self, *args, **options):
+        """Runs the accession gathering process."""
+        self.validate_args(options)
+        self.set_verbosity_level(options)
+
+        agents = list()
+        if options["rna_seq"]:
+            agents.append(RNASeqAccessionAgent(options))
+        elif options["microarray"]:
+            if (
+                options["ae_id"]
+                or options["ae_ids_file"]
+                or options["keyword"]
+                or options["organism"]
+            ):
+                agents.append(MicroArrayExpressAccessionAgent(options))
+            if (
+                options["gpl_id"]
+                or options["gpl_ids_file"]
+                or options["keyword"]
+                or options["organism"]
+            ):
+                agents.append(MicroArrayGEOAccessionAgent(options))
+
+        entries = set()
+        for agent in agents:
+            entries.update(agent.collect_data())
+
+        entries = sorted(  # Sort the resulting list.
+            (entry for entry in entries if self.RE_ACCESSION.match(entry.code)),
+            key=lambda entry: (
+                self.RE_ACCESSION.match(entry.code).group(1),
+                int(self.RE_ACCESSION.match(entry.code).group(2)),
+            ),
+        )
+        # Limit the number of output entries.
+        entries = entries[: options["count"]] if options["count"] else entries
+
+        if options["dry_run"]:
+            if entries:
+                output = "\n".join((str(entry) for entry in entries))
+            else:
+                output = "No accessions found."
+            print(output)
+        else:
+            AccessionBacklogEntry.objects.bulk_create(entries)
+
+
+class AccessionAgentBase:
```

Comment:
Can we pull these out of the command so that it is easier to see what the command does at a glance?